courses/models.py

```diff
@@ -0,0 +1,119 @@
+from django.db import models
+from utils.constants import SKILL_BEGINNER, SKILL_INTERMEDIATE, SKILL_ADVANCED, EXTERNAL_LINK_BLOG, EXTERNAL_LINK_TUTORIAL, EXTERNAL_LINK_COURSE
+from django.utils.text import slugify
+# Create your models here.
+
+class Language(models.Model):
+    language_name = models.CharField(max_length=20)
+    slug = models.SlugField(unique=True)
+
+    def _get_unique_slug(self):
+        slug = slugify(self.language_name)
+        unique_slug = slug
+        num = 1
+        while Language.objects.filter(slug=unique_slug).exists():
+            unique_slug = '{}-{}'.format(slug, num)
+            num += 1
+        return unique_slug
+
+    def save(self, *args, **kwargs):
```

Comment:
```py
from django.template.defaultfilters import slugify
from django.db import IntegrityError
from django.utils.crypto import get_random_string


def custom_slugify(source_field, suffix=False):
    """
    Using django util methods create a slug.
    Append a random string at the end of the slug if necessary for making it unique
    """
    new_slug = slugify(source_field)

    if suffix:
        random_str = get_random_string(length=10)
        new_slug = "{0}-{1}".format(new_slug, random_str)

    return new_slug


class Language(models.Model):
    language_name = models.CharField(max_length=20)
    slug = models.SlugField(unique=True)

    def save(self, *args, **kwargs):
        if not self.slug:
            self.slug = custom_slugify(source_field=self.language_name, suffix=False)
        """
        PLEASE TEST THIS CODE. I DID NOT GET A CHANCE TO TEST IT MYSELF.

         THE KEY TAKEAWAY FROM THE IS CODE PIECE IS THAT YOU NEED TO USE BUILT IN DJANGO CODE WHEREVER NECESSARY
         AS THIS PREVENT THE HEADACHE OF CODE MAINTAINANCE AND REDUCES TESTING
        """

        try:
            """
            While saving, it we have a duplicate combination of name and slug
            Since we have added an IntegrityError check, we will have an exception raised
            """
            super(Language, self).save(*args, **kwargs)
        except IntegrityError:
            """
            after catching the exception, we will generate another slug but this time we will
            suffix  it with a random string at the end of the slug to make it unique and try saving it again
            """
            self.slug = custom_slugify(source_field=self.language_name, suffix=True)
            super(Language, self).save(*args, **kwargs)

    def __str__(self):
        return '{name}-{slug}'.format(name=self.language_name, slug=self.slug)

    class Meta:
        unique_together = ('language_name', 'slug')
```